src/backgroundblurvideoframeprocessor/BackgroundBlurProcessorProvided.ts

```diff
@@ -0,0 +1,413 @@
+    if (!this.promise) {
+      // externally-resolvable promise
+      this.promise = new Promise(resolve => (this.resolve = resolve));
+    }
+    return this.promise;
+  }
+
+  /** Update the value and resolve */
+  next(value: T): void {
+    // store the value, for sync access
+    this.value = value;
+    // resolve the promise so anyone awaiting whenNext resolves
+    this.resolve(value);
+    // delete the promise so future whenNext calls get a new promise
+    delete this.promise;
+  }
+}
+
+/**
+ * [[BackgroundBlurVideoFrameProcessor]] implements [[VideoFrameProcessor]].
+ * It's a background blur processor and input is passed into a worker that will apply a segmentation
+ * to separate the foreground from the background. Then the background will have a blur applied.
+ */
+/** @internal */
+export default class BackgroundBlurProcessorProvided implements BackgroundBlurProcessor {
+  protected targetCanvas: HTMLCanvasElement = document.createElement('canvas') as HTMLCanvasElement;
+  protected canvasCtx = this.targetCanvas.getContext('2d');
+  protected canvasVideoFrameBuffer = new CanvasVideoFrameBuffer(this.targetCanvas);
+  protected mask$ = new DeferredObservable<ImageData>();
+  protected sourceWidth = 0;
+  protected sourceHeight = 0;
+  protected blurAmount = 0;
+
+  protected spec: BackgroundFilterSpec;
+  protected _blurStrength: number;
+  private delegate: BackgroundBlurVideoFrameProcessorDelegate;
+  protected worker: Worker;
+  protected logger: Logger;
+
+  protected initWorkerPromise = BackgroundBlurProcessorProvided.createWorkerPromise<{
+    message?: string;
+  }>();
+  protected loadModelPromise = BackgroundBlurProcessorProvided.createWorkerPromise<{
+    message?: string;
+  }>();
+  protected scaledCanvas: HTMLCanvasElement;
+  private frameCounter: BackgroundFilterFrameCounter;
+  protected modelInitialized = false;
+
+  private static createWorkerPromise<T>(): {
+    resolve: (value: T) => void;
+    reject: (value: Error) => void;
+    promise: Promise<T>;
+  } {
+    const resolver: {
+      resolve: (value: T) => void;
+      reject: (value: Error) => void;
+      promise: Promise<T>;
+    } = { resolve: null, reject: null, promise: null };
+    resolver.promise = new Promise<T>((resolve, reject) => {
+      resolver.resolve = resolve;
+      resolver.reject = reject;
+    });
+    return resolver;
+  }
+
+  private validateSpec(spec: BackgroundFilterSpec): void {
+    if (!spec) {
+      throw new Error('processor has null spec');
+    }
+
+    if (!spec.model) {
+      throw new Error('processor spec has null model');
+    }
+
+    if (!spec.paths) {
+      throw new Error('processor spec has null paths');
+    }
+  }
+
+  private validateOptions(options: BackgroundBlurOptions): void {
+    if (!options) {
+      throw new Error('processor has null options');
+    }
+
+    if (!options.blurStrength) {
+      throw new Error('processor has null options - blurStrength');
+    }
+
+    if (!options.logger) {
```

Comment:
Why make options optional if we are going to throw exceptions if the aren't set :P.